add warning to kubectl auth can-i if a subresource does not exist

atiratree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Adds a warning if a subresource is specified in `kubectl auth can-i ` that does not exist, but still preserves API behaviour and outputing result from SubjectAccessReview

Prefer builtin non-discoverable resources instead of custom resources (users, groups) when versionGroup is not specified.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubectl/issues/1217

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
TBD after reviews

```release-note
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

atiratree committed:
add warning to kubectl auth can-i if a subresource does not exist

prefer builtin non-discoverable resources instead of custom resources
(users, groups) when versionGroup is not specified

atiratree commented:
/sig cli

soltysh commented:
/assign @ardaguclu @deejross

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/auth/cani.go:
```diff
@@ -325,6 +336,13 @@ func (o *CanIOptions) resourceFor(mapper meta.RESTMapper, resourceArg string) sc
 		}
 	}
 
+	if len(subresourceArg) > 0 && o.DiscoveryClient != nil {
```
We may remove this null check, because we are failing above if discoveryClient gets error

atiratree replied:
this is so the tests pass - I guess we could use a mock discovery there. Also I still need to work on the tests

liggitt replied:
if we require a non-nil discovery client, then checking for nil here and taking different branches doesn't seem great

atiratree replied:
I made sure to always supply the discovery client in the tests, so the null check is not necessary anymore.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/auth/cani.go:
```diff
@@ -416,6 +434,38 @@ func isNamespaced(gvr schema.GroupVersionResource, discoveryClient discovery.Dis
 	return false, fmt.Errorf("the server doesn't have a resource type '%s' in group '%s'", gvr.Resource, gvr.Group)
 }
 
+func hasSubresource(discoveryClient discovery.DiscoveryInterface, gvr schema.GroupVersionResource, subresource string) error {
```
Wouldn't be better and easier to do that check in api part instead in here?. This looks to me a complex effort in client side to verify the existence of subresource. Moreover it is a low probability but subresource may not be found due to 6 hours TTL?

Thoughts?

atiratree replied:
SARs are checking only the matching against the rules (RBAC) not existince of a subresource

atiratree commented on staging/src/k8s.io/kubectl/pkg/cmd/auth/cani.go:
```diff
@@ -169,7 +169,12 @@ func (o *CanIOptions) Complete(f cmdutil.Factory, args []string) error {
 				return err
 			}
-			o.Resource = o.resourceFor(restMapper, resourceTokens[0])
+			discoveryClient, err := f.ToDiscoveryClient()
```
to have it in resourceFor. I guess we could do it in the beginning of the function - I just wanted to make the execution leaner in case we give an error to the user before that

liggitt commented:
I haven't reviewed the code yet, but I want to make sure this is not introducing something that will fail a command that previously worked

`can-i` should be able to interrogate the authorization system for permissions even for resources which are not discoverable

atiratree commented:
> I haven't reviewed the code yet, but I want to make sure this is not introducing something that will fail a command that previously worked
>
> can-i should be able to interrogate the authorization system for permissions even for resources which are not discoverable

@liggitt  this PR is trying to preserve this. I am adding only a warning on top of the original response

removed WIP and added tests

Dentrax commented:
Kind ping 🤞

enj commented:
/assign @nilekhc 
(for review)

nilekhc commented:
/lgtm

pending other comments.

nilekhc commented:
This LGTM

/cc @enj @liggitt PTAL when you get a chance.

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/auth/cani.go:
```diff
@@ -325,6 +336,13 @@ func (o *CanIOptions) resourceFor(mapper meta.RESTMapper, resourceArg string) sc
 		}
```
I don't see this branch considering the subresource

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/auth/cani.go:
```diff
@@ -416,6 +434,40 @@ func isNamespaced(gvr schema.GroupVersionResource, discoveryClient discovery.Dis
+		}
+	}
+	if serverResources == nil {
+		_, serverResources, _ = discoveryClient.ServerGroupsAndResources()
```
this has the potential to be an expensive call on a server with many API groups/versions... I wouldn't fall back to this just for a warning

atiratree replied:
`kubectl` should cache these, but yes if the cache is out it could trigger an expensive call. I have removed the fallback as it does not seem necessary now.

atiratree committed:
add tests for kubectl auth can-i subresource warning

atiratree commented on staging/src/k8s.io/kubectl/pkg/cmd/auth/cani.go:
```diff
@@ -309,28 +356,64 @@ func (o *CanIOptions) RunAccessCheck() (bool, error) {
+	}
+
+	// try to match a non-standard resource first
+	gvr, err := gvrNonStandardResourceSupplier(resourceArg, subresourceArg)
```
I think we should try to resolve non standard resources first before we match them via a mapper. So for example we will choose `users` before e.g. `users.user.openshift.io`. This is a breaking change, but IMO it makes more sense to prefer the builtins. It certainly makes more sense for the [impersonate](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation) verb. I suppose we could use builtins only for this verb and use custom resources for the other verbs, but I do not like the inconsistency we might introduce. If people want to check for their resources they can always use fully qualified names.

@liggitt thoughts?

liggitt replied:
I'm actually pretty skeptical about doing all this work and surfacing way more warnings in general. The authorization system and API is intentionally opaque ... making the kubectl command "smarter" gets in the way of using the authorization system that way, and hard-coding in exceptions doesn't work well in the face of client/server skew.

Is there evidence that these additional warnings are needed, and the benefits outweigh the increased friction for people using this command to check synthetic permissions?

atiratree replied:
I think we should first discuss the resource resolution before the we decide on the warnings. At least a couple of users think these warnings are useful, but I have no other data to back that up. The warnings are just an add-on to and do not change the authorization system.

What changes the authorization system is the resource resolution according to the the mapper/discovery. It may have different results in case user just specifies `resource` instead of `resource.version.group` or `resource.group`. I am not sure we can backtrack on this (make it less smart) and force users to always fully specify the resource, as that would break the command backwards compatibility.  In this respect, I think we have a bug as it is not possible to check for the non standard resources (`users`, and the other 3) at the moment, and this is the main reason why the PR is more complicated.

I do not like the hard coding of the values, but it seems like an okay solution at this time to me. I suppose we could include this in the discovery or alternate mechanism. Although that seems out of scope for this PR. Do you think it makes sense to propose such a feature?

atiratree replied:
@liggitt ^

atiratree reviewed (commented):
I had to rewrite the code a bit and the scope includes better handling of non standard resources now.

atiratree commented:
/retest
